File: sdk-ts/src/authentication/clientcredentials.ts
import { saveCredentials } from "./credentials";
import { Credentials } from "./types";

interface DeviceLoginFinalizeResponse {
  access_token: string;
  expires_in: number;
  refresh_token: string;
  token_type: string;
}

export class ClientCredentials {
  private credentials: Credentials;
  private workspace_name: string;
  private base_url: string;

  constructor(
    credentials: Credentials,
    workspace_name: string,
    base_url: string
  ) {
    this.credentials = credentials;
    this.workspace_name = workspace_name;
    this.base_url = base_url;
  }

  async getHeaders(): Promise<Record<string, string>> {
    const err = await this.refreshIfNeeded();
    if (err) {
      throw err;
    }

    return {
      "X-Beamlit-Authorization": `Bearer ${this.credentials.access_token}`,
      "X-Beamlit-Workspace": this.workspace_name,
    };
  }

  async refreshIfNeeded(): Promise<null> {
    if (
      this.credentials.client_credentials &&
      !this.credentials.refresh_token
    ) {
      const body = { grant_type: "client_credentials" };

      try {
        const response = await fetch(`${this.base_url}/oauth/token`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Basic ${this.credentials.client_credentials}`,
          },
          body: JSON.stringify(body),
        });
        const data = (await response.json()) as DeviceLoginFinalizeResponse;
        this.credentials.access_token = data.access_token;
        this.credentials.refresh_token = data.refresh_token;
        this.credentials.expires_in = data.expires_in;
      } catch (e) {
        throw new Error(`Failed to get client credentials: ${e}`);
      }
    }

    // Need to refresh token if expires in less than 10 minutes
    const parts = this.credentials.access_token?.split(".") || [];
    if (parts.length !== 3) {
      throw new Error("Invalid JWT token format");
    }

    try {
      const claims = JSON.parse(
        Buffer.from(parts[1], "base64").toString("utf-8")
      );
      const expTime = claims.exp * 1000; // Convert to milliseconds
      // Refresh if token expires in less than 10 minutes
      if (Date.now() + 10 * 60 * 1000 > expTime) {
        return await this.doRefresh();
      }
    } catch (e) {
      throw new Error(`Failed to decode/parse JWT claims: ${e}`);
    }

    return null;
  }

  intercept(req: Request): void {
    const err = this.refreshIfNeeded();
    if (err) {
      throw err;
    }

    req.headers.set(
      "X-Beamlit-Authorization",
      `Bearer ${this.credentials.access_token}`
    );
    req.headers.set("X-Beamlit-Workspace", this.workspace_name);
  }

  private async doRefresh(): Promise<null> {
    if (!this.credentials.refresh_token) {
      throw new Error("No refresh token to refresh");
    }

    const url = `${this.base_url}/oauth/token`;
    const refresh_data = {
      grant_type: "refresh_token",
      refresh_token: this.credentials.refresh_token,
      device_code: this.credentials.device_code,
      client_id: "beamlit",
    };

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(refresh_data),
      });

      const finalize_response =
        (await response.json()) as DeviceLoginFinalizeResponse;

      if (!finalize_response.refresh_token) {
        finalize_response.refresh_token = this.credentials.refresh_token;
      }

      const creds: Credentials = {
        access_token: finalize_response.access_token,
        refresh_token: finalize_response.refresh_token,
        expires_in: finalize_response.expires_in,
        device_code: this.credentials.device_code,
        client_credentials: this.credentials.client_credentials,
      };

      this.credentials = creds;
      saveCredentials(this.workspace_name, creds);
      return null;
    } catch (e) {
      throw new Error(`Failed to refresh token: ${e}`);
    }
  }
}

File: sdk/clientcredentials.go
package sdk

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type ClientCredentials struct {
	credentials   Credentials
	workspaceName string
	baseUrl       string
}

func NewClientCredentialsProvider(credentials Credentials, workspaceName string, baseUrl string) *ClientCredentials {
	return &ClientCredentials{credentials: credentials, workspaceName: workspaceName, baseUrl: baseUrl}
}

func (c *ClientCredentials) GetCredentials() Credentials {
	return c.credentials
}

func (c *ClientCredentials) RefreshIfNeeded() error {
	// Handle client credentials flow if needed
	if c.credentials.ClientCredentials != "" && c.credentials.RefreshToken == "" {
		headers := map[string]string{
			"Authorization": fmt.Sprintf("Basic %s", c.credentials.ClientCredentials),
			"Content-Type":  "application/json",
		}

		body := map[string]string{
			"grant_type": "client_credentials",
		}

		jsonData, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("failed to marshal client credentials data: %v", err)
		}

		req, err := http.NewRequest("POST", c.baseUrl+"/oauth/token", strings.NewReader(string(jsonData)))
		if err != nil {
			return fmt.Errorf("failed to create request: %v", err)
		}

		for k, v := range headers {
			req.Header.Add(k, v)
		}

		res, err := http.DefaultClient.Do(req)
		if err != nil {
			return fmt.Errorf("failed to execute request: %v", err)
		}
		defer res.Body.Close()

		var response DeviceLoginFinalizeResponse
		if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
			return fmt.Errorf("failed to decode response: %v", err)
		}

		c.credentials.AccessToken = response.AccessToken
		c.credentials.RefreshToken = response.RefreshToken
		c.credentials.ExpiresIn = response.ExpiresIn

		return nil
	}

	// Need to refresh the token if access token expires in less than 10 minutes
	parts := strings.Split(c.credentials.AccessToken, ".")
	if len(parts) != 3 {
		return fmt.Errorf("invalid JWT token format")
	}

	claimsBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return fmt.Errorf("failed to decode JWT claims: %v", err)
	}

	var claims map[string]interface{}
	if err := json.Unmarshal(claimsBytes, &claims); err != nil {
		return fmt.Errorf("failed to parse JWT claims: %v", err)
	}

	expTime := time.Unix(int64(claims["exp"].(float64)), 0)
	// Refresh if token expires in less than 10 minutes
	if time.Now().Add(10 * time.Minute).After(expTime) {
		return c.DoRefresh()
	}

	return nil
}

func (c *ClientCredentials) Intercept(ctx context.Context, req *http.Request) error {
	if err := c.RefreshIfNeeded(); err != nil {
		return err
	}

	req.Header.Set("X-Blaxel-Authorization", fmt.Sprintf("Bearer %s", c.credentials.AccessToken))
	req.Header.Set("X-Blaxel-Workspace", c.workspaceName)
	return nil
}

func (c *ClientCredentials) DoRefresh() error {
	if c.credentials.RefreshToken == "" {
		return fmt.Errorf("no refresh token to refresh")
	}
	url := c.baseUrl + "/oauth/token"
	refreshData := map[string]string{
		"grant_type":    "refresh_token",
		"refresh_token": c.credentials.RefreshToken,
		"device_code":   c.credentials.DeviceCode,
		"client_id":     "blaxel",
	}

	jsonData, err := json.Marshal(refreshData)
	if err != nil {
		return fmt.Errorf("failed to marshal refresh data: %v", err)
	}

	payload := strings.NewReader(string(jsonData))

	req, _ := http.NewRequest("POST", url, payload)

	req.Header.Add("content-type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to refresh token: %v", err)
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(res.Body)

	var finalizeResponse DeviceLoginFinalizeResponse
	if err := json.Unmarshal(body, &finalizeResponse); err != nil {
		panic(err)
	}
	if finalizeResponse.RefreshToken == "" {
		finalizeResponse.RefreshToken = c.credentials.RefreshToken
	}
	creds := Credentials{
		AccessToken:  finalizeResponse.AccessToken,
		RefreshToken: finalizeResponse.RefreshToken,
		ExpiresIn:    finalizeResponse.ExpiresIn,
		DeviceCode:   c.credentials.DeviceCode,
	}

	c.credentials = creds

	SaveCredentials(c.workspaceName, creds)

	return nil
}
